HARK/ConsumptionSaving/ConsRiskyContribModel.py

```diff
@@ -0,0 +1,1768 @@
+        self.state_now["nNrmTilde"] = nNrmTilde
+
+    def get_controls_Sha(self):
+        """
+        Get controls for the second "stage": choosing the contribution share.
+        """
+
+        Share = np.zeros(self.AgentCount) + np.nan
+
+        # Loop over each period of the cycle, getting controls separately depending on "age"
+        for t in range(self.T_cycle):
+
+            # Find agents in this period-stage
+            these = t == self.t_cycle
+
+            # Get controls for agents who *can* adjust.
+            those = np.logical_and(these, self.shocks["Adjust"])
+            Share[those] = (
+                self.solution[t]
+                .stage_sols["Sha"]
+                .ShareFunc_Adj(
+                    self.state_now["mNrmTilde"][those],
+                    self.state_now["nNrmTilde"][those],
+                )
+            )
+
+            # Get Controls for agents who *can't* adjust.
+            those = np.logical_and(these, np.logical_not(self.shocks["Adjust"]))
+            Share[those] = (
+                self.solution[t]
+                .stage_sols["Sha"]
+                .ShareFunc_Fxd(
+                    self.state_now["mNrmTilde"][those],
+                    self.state_now["nNrmTilde"][those],
+                    self.state_prev["Share"][those],
+                )
+            )
+
+        # Store controls as attributes of self
+        self.controls["Share"] = Share
+
+    def get_states_Cns(self):
+        """
+        Get states for the third "stage": consumption.
+        """
+
+        # Contribution share becomes a state in the consumption problem
+        self.state_now["Share"] = self.controls["Share"]
+
+    def get_controls_Cns(self):
+        """
+        Get controls for the third "stage": consumption.
+        """
+
+        cNrm = np.zeros(self.AgentCount) + np.nan
+
+        # Loop over each period of the cycle, getting controls separately depending on "age"
+        for t in range(self.T_cycle):
+
+            # Find agents in this period-stage
+            these = t == self.t_cycle
+
+            # Get consumption
+            cNrm[these] = (
+                self.solution[t]
+                .stage_sols["Cns"]
+                .cFunc(
+                    self.state_now["mNrmTilde"][these],
+                    self.state_now["nNrmTilde"][these],
+                    self.state_now["Share"][these],
+                )
+            )
+
+        # Store controls as attributes of self
+        # Since agents might be willing to end the period with a = 0, make
+        # sure consumption does not go over m because of some numerical error.
+        self.controls["cNrm"] = np.minimum(cNrm, self.state_now["mNrmTilde"])
+
+    def get_post_states(self):
+        """
+        Set variables that are not a state to any problem but need to be
+        computed in order to interact with shocks and produce next period's
+        states.
+        """
```

Comment:
Interesting that the motivation for the post_state architecture seems to mainly be Matt White's 2015 paper on EGM.

I feel like that should be better documented. (would have saved me a lot of time) probably an aside not appropriate for this PR review....

Reply:
Yes. I do not mean to resuscitate a dead horse but while working on this I gained a renewed appreciation for post-states. 

They are useful for both solutions and simulations. Maybe more so as HARK shifts towards transition equations and more modular representations of models. At the moment, I am thinking of post-states as "The minimal set of variables that, coupled with the joint distribution of shocks, can be used to produce the joint distribution of next-period's states". In this model's case those would be `[aNrm, nTildeNrm]`, but `nTildeNrm` is already a state.

@llorracc told me that his vision of solvers includes splitting off the expectation-formation part (where one computes E[v(m)], E[v'(m)], etc). Under this previous definition, post-states would be precisely the numbers that this expectation-calculator would need to know (more precisely, have grids over, since these expectations would be functions of post-states).